eval/adjust.rkt

```diff
@@ -71,333 +71,263 @@
-    
+     (list (+ (- (maxlog x) (minlog z)) slack))]
+
     ; Currently expm1 has a very poor solution for negative values
     [(ival-expm1)
-     ; log[Гexpm1] = log[x * e^x / expm1] <= max(1+log[x], 1+log[x/expm1] +1)
-     ;                                                                    ^^ division accounting
+     ; log[Гexpm1] = log[x * e^x / expm1] <= max(1 + maxlog(x), 1 + maxlog(x) - minlog(z))
      (define x (first srcs))
-     (define x-exp (ival-max-log2-approx x))
-     (define out-exp (ival-min-log2-approx output))
-     
-     (list (max (+ 1 x-exp) (+ 2 (- x-exp out-exp))))]
+     (list (max (+ 1 (maxlog x)) (+ 1 (- (maxlog x) (minlog z)))))]
     
     [(ival-atan2)
-     ; log[Гatan2]'x = log[Гatan2]'y = log[xy / ((x^2+y^2)*atan2)] <= log[x] + log[y] - 2*max[logx, logy] - log[atan2] + 1
+     ; maxlog(x) + maxlog(y) - 2*max(minlog(x), minlog(y)) - minlog(z)
      (define x (first srcs))
      (define y (second srcs))
      
-     (define x-exp (ival-max-log2-approx x))
-     (define y-exp (ival-max-log2-approx y))
-     (define out-exp (ival-min-log2-approx output))
-     
-     (make-list 2 (+ (- (+ x-exp y-exp) (* 2 (max x-exp y-exp)) out-exp) 1))]
+     (make-list 2 (- (+ (maxlog x) (maxlog y)) (* 2 (max (minlog x) (minlog y))) (minlog z)))]
     
-    ; Currently has a poor implementation
     [(ival-tanh)
-     ; log[Гtanh] = log[x / (sinh(x) * cosh(x))] <= -log[x] + log[tanh]. Never greater than 0
+     ; logspan(z) + logspan(x)
      (define x (first srcs))
-     (define x-exp (ival-min-log2-approx x))
-     (define out-exp (ival-max-log2-approx output))
-     
-     (list (+ (- x-exp) out-exp))]
-    
+     (list (+ (logspan z) (logspan x)))]
+
     [(ival-atanh)
-     ; log[Гarctanh] = log[x / ((1-x^2) * atanh)] = 1 if x < 0.5, otherwise slack
-     ;                          ^^^^^^^
-     ;                          a possible uncertainty
+     ; log[Гarctanh] = maxlog(x) - log[(1-x^2)] - minlog(z) = 1 if x < 0.5, otherwise slack
+     ;                               ^^^^^^^
+     ;                           a possible uncertainty
      (define x (first srcs))
-     (define x-exp (ival-max-log2-approx x))
-     
-     (list (if (>= x-exp 1)
+     (list (if (>= (maxlog x) 1)
                (get-slack)
                1))]
-    
+
     [(ival-acosh)
-     ; log[Гacosh] = log[x / (sqrt(x-1) * sqrt(x+1) * acosh)] <= -log[acosh] + slack
-     (define out-exp (ival-min-log2-approx output))
-     (define slack (if (< out-exp 2)                  ; when acosh(x) < 1
+     ; log[Гacosh] = log[x / (sqrt(x-1) * sqrt(x+1) * acosh)] <= -minlog(z) + slack
+     (define z-exp (minlog z))
+     (define slack (if (< z-exp 2)                    ; when acosh(x) < 1
                        (get-slack)
                        0))
      
-     (list (+ (- out-exp) slack))]
+     (list (- slack z-exp))]
+
+    [(ival-pow2)
+     ; same as multiplication
+     (define x (first srcs))
+     (list (logspan x))]
```

Comment:
I think it's actually `(logspan x) + 1` because there are two arguments and they are the same, basically.

Reply:
Added